clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp

```diff
@@ -87,6 +85,28 @@ bool isStdVariant(const Type *Type) {
   return isStdType(Type, llvm::StringLiteral("variant"));
 }
 
+bool isStdAny(const Type *Type) {
+  return isStdType(Type, llvm::StringLiteral("any"));
+}
+
+bool isVowel(char a) {
+  switch (a) {
+  case 'a':
+  case 'e':
+  case 'i':
+  case 'o':
+  case 'u':
+    return true;
+  default:
+    return false;
+  }
+}
+
+llvm::StringRef indefiniteArticleBasedOnVowel(char a) {
+  if (isVowel(a))
+    return "an";
+  return "a";
+}
```

Comment:
I don't think this is phonetically correct. For example, if I call this for the first letter in the word `union`, which is pronounced as [`YUU-ni-yon` or `YUUN-jen`](http://en.wiktionary.org/wiki/union#English), then it should be **`a union`** and not `an union`.

Some more examples: http://english.stackexchange.com/questions/266309/why-is-union-an-exception-to-the-a-an-rule

Might be worthwhile to just get rid of this altogether, and refactor the call site like this: `any held a value of 'Union'` or `any contained object of type 'Union'`.

Reply:
I like the new sentence you suggested. I will replace the current warning text for this checker and for `std::variant` checker too.

Reply:
I'm bikeshedding this in the other direction, because I think that the length / verbosity of the messages is an important factor, so we shouldn't insert meaningless words just to simplify the code. We do not need absolutely perfect grammar, if we're attaching an article to an arbitrary variable name, it's OK if there are rare situations when it doesn't fit the pronunciation.

Reply:
I think a good solution would be just to use the current text, but instead of deciding the indefinite article dynamically, we could just use `a(n)`. I think with this solution we would have correct grammar and a short warning. What do you think?

Reply:
I think parenthesized `a(n)` in 100% of the messages is uglier than incorrectly applied `a` or `an` in <1% of the messages (in average English text, 1% of the words start with `u` [1] -- this may be different for type names, but the order of magnitude should be similar).

However, this is just bikesheddig, so I won't block any of the three solutions (decision based on `isVowel()`, using `a(n)` and using filler words to dodge the issue).

[1] https://www.johndcook.com/blog/2021/08/16/initial-letter-frequency/

Reply:
> I think that the length / verbosity of the messages is an important factor, so we shouldn't insert meaningless words just to simplify the code.

Actually, I believe @DonatNagyE is right. There isn't much issue in saying `std::any contained 'Union'`. No real need for articles here.

Reply:
If you skip the article, the message looks as if it was reporting the _value_ stored in the `std::any`, which may be an annoying confusion if the type name is not immediately recognizable as a type name.

I see that it has some advantages, but overall I think it's worse than any of the three other solutions (decision based on `isVowel()`, using a(n) and using filler words to dodge the issue).